### PR TITLE
Subtitles

### PR DESCRIPTION
**Pull request by kirill09**
subtitle display

### CODE REVIEW

**Comment by Reached** [comment]
@kirill09 is this working for you? :)

**Comment by kirill09** [comment]
> @kirill09 is this working for you? :)

yes

**Comment by Ahmadre** [comment]
@kirill09 thank you sooo much for this PR :) whats about the conflicts? could you check them?

**Comment by Ahmadre** [comment]
@kirill09 I think these changes from you are very important and give chewie a big plus. I would appreciate it, if you have a second look here into this PR and the merge conflicts, so we can have subtitles in chewie :)

**Comment by Ahmadre** [comment]
I fixed the conflicts. I will check everything now. @nstrelow if you got time you can also have a quick look. Maybe I am too fast 😂

**Comment by Ahmadre** [comment]
Ok, tested everything, here are my results:
I don't understand why we're using such dirty algorithm to pass:
- index
- start duration
- end duration
- text
by this:
```dart
subtitle: Subtitles([
    Subtitle('0\n00:00:00 --> 00:00:10\nHello from Subtitles :)'),
]),
subtitleBuilder: (context, subtitle) {
    return Text(subtitle);
},
```

The result in the UI is:
Bildschirmfoto 2020-11-12 um 20 29 14

Honestly this is not really user or developer friendly. It's very easy in Dart to export a simple model for this. Yes, model is existing but factory constructor is horrible ^^ sorry to say that.

Also: the parameter: subtitle was existing as a parameter but was not passed into the constructor, so I had to fix that.

My suggestion: I'll refactor the Subtitle model to pass user-friendly parameters rather then a string like above.

/cc @nstrelow

File: lib/src/chewie_player.dart [hunk 1]
```diff
@@ -269,6 +270,8 @@ class ChewieController extends ChangeNotifier {
   /// Defines a custom RoutePageBuilder for the fullscreen
   final ChewieRoutePageBuilder routePageBuilder;
 
+  Subtitles subtitle;
```
Comment (Ahmadre) [inline lib/src/chewie_player.dart:322]: This wasn't mention in the constructor. Already fixed.

File: lib/src/subtitle_model.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  static const String timerSeparator = ' --> ';
+  final int index;
+  final Duration start;
+  final Duration end;
+  final List<String> texts;
+
+  static Duration stringToDuration(String value) {
+    final component = value.split(':');
+    return Duration(
```
Comment (Ahmadre) [inline lib/src/subtitle_model.dart]: Added check to also pass: 00:00:00 if milliseconds not needed. Already fixed.

File: lib/src/subtitle_model.dart [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    final found = subtitle.where(
+            (item) => position >= item.start && position <= item.end
+    ).toList();
+
+    return found;
+  }
+}
+
+class Subtitle {
+  factory Subtitle(String value) {
```
Comment (Ahmadre) [inline lib/src/subtitle_model.dart:53]: factory constructor will be refactored so it's more user friendly to use.

File: lib/src/cupertino_controls.dart [hunk 4]
```diff
@@ -75,6 +79,7 @@ class _CupertinoControlsState extends State<CupertinoControls> {
           children: <Widget>[
             _buildTopBar(backgroundColor, iconColor, barHeight, buttonPadding),
             _buildHitArea(),
+            _buildSubtitles(chewieController.subtitle),
```
Comment (Ahmadre) [inline lib/src/cupertino_controls.dart]: Doesn't align relatively to the video aspect ratio. Will be fixed.

**Comment by kennylbj** [comment]
Is this PR ready to merged?

**Comment by Ahmadre** [comment]
> Is this PR ready to merged?

Nope, I am working on it 🧑🏻‍💻

**Comment by Ahmadre** [comment]
Hey guys 👋

Sorry for the delay. I moved many times inside Germany and now I can continue my work on this important feature.

I already got it working, but it's hidden by the controls and I want to make the subtitles aware of the controls.

Be prepared 💪🏼

**Comment by Ahmadre** [comment]
Since @kirill09 's repo and branch are somehow damaged/deleted, I've moves this PR to: https://github.com/brianegan/chewie/pull/475

please follow that PR. It's ready to release, we just have to review it!

**Pull request closed without merging**